crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1243,6 +1257,35 @@ impl<'db> TypeInferenceBuilder<'db> {
         ty
     }
 
+    fn infer_annotation_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+        // https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-annotation_expression
+        let ty = match expression {
+            // type_expression is responsible for returning an error in the case
+            // that the name is somehow in an invalid context.
+            name @ ast::Expr::Name(..) => self.infer_type_expression(name),
+            // TODO Remaining forms from grammar
+            _ => Type::Unknown, // TODO: this needs to error?
+        };
+
+        ty
+    }
+    fn infer_type_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+        // https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-type_expression
+        // TODO: this does not include any of the special forms, and is only a
+        //   stub of the forms other than a standalone name in scope.
+        match expression {
+            ast::Expr::Name(name) => {
+                assert!(
+                    name.ctx.is_load(),
+                    "name in a type expression is always 'load' but got: '{:?}'",
+                    name.ctx
+                );
```

Comment:
Maybe make this a debug assertion? I think it's unlikely it'll ever trigger

```suggestion
                debug_assert!(
                    name.ctx.is_load(),
                    "name in a type expression is always 'load' but got: '{:?}'",
                    name.ctx
                );
```

Reply:
Seems reasonable—here I was following the example of the `panic!` on `infer_function_type_params` (in a `let...else`), though in that case it's necessary simply for using the construct.